### ISSUE TITLE
How to format version if the repository’s current state is an exact tag match

### ISSUE THREAD

**Issue by mzebrak** [jwodder/versioningit#42]
Is there an option to format the version like the `[tool.versioningit.format]` does, but when there is no distance and is no dirty?

I need to do something like:
```toml
[tool.versioningit.format]
base = "{base_version}+{vcs}{rev}.{author_date:%Y%m%dT%H%M%S}"
distance = "{base_version}.post{distance}+{vcs}{rev}.{author_date:%Y%m%dT%H%M%S}"
dirty = "{base_version}.dev+{vcs}{rev}.{author_date:%Y%m%dT%H%M%S}"
distance-dirty = "{base_version}.dev{distance}+{vcs}{rev}.{author_date:%Y%m%dT%H%M%S}"
```

**Comment by jwodder** [comment]
No, there is currently no such option.  May I ask why you want that?  If the current commit has a tag, that indicates that the version should equal the one in the tag; that's kind of the point of storing version information in tags.  Adding on a local version component to a released version string is just unusual; for one thing, you wouldn't be able to upload your project to PyPI, as it forbids local version components.

**Comment by mzebrak** [comment]
In the project in which I am involved, versioning works in the way I mentioned above. Unfortunately, I need to add additional information in addition to the tag itself. I found that versioningit has one of the best configuration options but I miss the option mentioned above. is there any workaround? I see you determine it early in `run()` function of the `core.py`, so any overriding of the `vcs` or `tag2version` will not work :/

**Comment by jwodder** [comment]
I believe the following would work as a workaround:

- Define a custom VCS method that acts like the `git` method but sets `description.state` to something other than `"exact"` (say, to `"base"`) when there's no distance or dirt
- Supply a format template string under `[tool.versioningit.format]` for this custom state (like you did in your initial comment)

As for the possibility of versioningit supporting this directly without the need for a workaround, I'm struggling to see the utility or motivation behind this use case.  If more developers from more projects express a desire to be able to do this sort of thing, I'll add it, but at the moment I don't feel that a single oddball is enough to change the version-calculation flow.

**Comment by mzebrak** [comment]
Thank you for the quick reply, I really appreciate it. I was thinking about what you wrote earlier and I think this versioning method is not the best idea since it closes the possibility of uploading project to PyPI.


In my case It all comes down to the base usecase which is the need to generate a file during the python build stage which would contain info about `version`, `git_revision_date` , `git_revision_sha`. 

I just noticed that by using your tool I have the ability to define a template for this and I can easily enter all this information into the generated file. In that case, I don't need to store that data directly in the `__version__` itself as I tried before.
